homeassistant/components/modbus/modbus.py

```diff
@@ -314,13 +376,120 @@ async def async_setup(self) -> bool:
         self._async_cancel_listener = None
         self._config_delay = 0
+        self.start_update_listener()
+
+    def start_update_listener(self) -> None:
+        """Possibly start monitoring of updates."""
+        for (scan_group, interval_millis) in self._scan_groups.items():
+            _LOGGER.debug(
+                "Register scan listener scan_group=%s, interval_millis=%s",
+                scan_group,
+                interval_millis,
+            )
+            async_track_time_interval(
+                self.hass,
+                self.async_update_function(scan_group),
+                timedelta(milliseconds=interval_millis),
+            )
+
+    def register_update_listener(
+        self,
+        scan_group: str,
+        slave: int,
+        input_type: str,
+        min_address: int,
+        max_address: int,
+        func: Callable[
+            [ModbusResponse | None, int, str, int], Coroutine[Any, Any, None]
+        ],
+    ) -> None:
+        """Register update listener."""
+        _LOGGER.debug(
+            "Register update listener slave=%s, input_type=%s, min_address=%s, max_address=%s in scan_group=%s",
+            slave,
+            input_type,
+            min_address,
+            max_address,
+            scan_group,
+        )
+        update_listeners = self._update_listeners_by_scan_group[scan_group]
+        key = (slave, input_type)
+        if key in update_listeners:
+            update_listeners[key].append(
+                ModbusUpdateListener(slave, input_type, min_address, max_address, func)
+            )
+        else:
+            update_listeners[key] = [
+                ModbusUpdateListener(slave, input_type, min_address, max_address, func)
+            ]
+
+    def async_update_function(
+        self, scan_group: str
+    ) -> Callable[..., Awaitable[None] | None]:
+        """Return async update function per scan group."""
+
+        @callback
+        async def async_update(now: datetime | None = None) -> None:
+            """Update the state of all entities in a given scan group."""
+            # remark "now" is a dummy parameter to avoid problems with
+            # async_track_time_interval
+            for (
+                (slave, input_type),
+                listeners,
+            ) in self._update_listeners_by_scan_group[scan_group].items():
+                min_address = 100000
+                max_address = 0
+                for listener in listeners:
+                    min_address = min(min_address, listener._min_address)
+                    max_address = max(max_address, listener._max_address)
+                _LOGGER.debug(
+                    "query modbus: scan_group=%s, slave=%s, minAdress=%s, maxAdress=%s, input_type=%s",
+                    scan_group,
+                    slave,
+                    min_address,
+                    max_address,
+                    input_type,
+                )
+                result = await self.async_pymodbus_call(
```

Comment:
@janiversen 
Here we do the effectice modbus call, again independant of devices, just by input_type and per scan_group. Afterwards we update all registered devices with the starting_offset so the device knows where to start reading the results from it.
Why do you think listeners are expensive? Maybe it's just a wording issue, I meant listeners as normally used in java environments which means a reference to an object having a specific interface to be able to notify it. Is there a better way to inform a certain list of devices in pyton/HA?